pkg/schema/definitions/converter.go

```diff
@@ -0,0 +1,223 @@
+	}
+
+	def := definition{
+		Type:           path.String(),
+		Description:    props.Description,
+		ResourceFields: map[string]definitionField{},
+	}
+
+	requiredSet := make(map[string]struct{})
+	for _, name := range props.Required {
+		requiredSet[name] = struct{}{}
+	}
+
+	for name, prop := range props.Properties {
+		_, required := requiredSet[name]
+		field := convertJSONSchemaPropsToDefinitionField(prop, path.FieldPath(name), required)
+		def.ResourceFields[name] = field
+
+		convertJSONSchemaPropsToDefinition(prop, path.FieldPath(name), definitions)
+	}
+	definitions[path.String()] = def
+}
+
+func convertJSONSchemaPropsToDefinitionField(props apiextv1.JSONSchemaProps, path proto.Path, required bool) definitionField {
+	field := definitionField{
+		Description: props.Description,
+		Required:    required,
+		Type:        getPrimitiveType(props.Type),
+	}
+	switch props.Type {
+	case "array":
+		field.Type = "array"
+		if item := getItemsSchema(props); item != nil {
+			if item.Type == "object" || item.Type == "array" {
+				field.SubType = path.String()
+			} else {
+				field.SubType = getPrimitiveType(item.Type)
+			}
+		}
+	case "object":
+		field.Type = path.String()
+	}
+	return field
+}
+
+func getPrimitiveType(typ string) string {
+	switch typ {
+	case "string":
+		return "string"
+	case "boolean":
+		return "boolean"
+	case "integer", "number":
+		return "int"
+	}
+	return ""
+}
+
+func getItemsSchema(props apiextv1.JSONSchemaProps) *apiextv1.JSONSchemaProps {
+	if props.Items == nil {
+		return nil
+	}
+
+	if props.Items.Schema != nil {
+		return props.Items.Schema
+	} else if len(props.Items.JSONSchemas) > 0 {
+		return &props.Items.JSONSchemas[0]
+	}
+	return nil
+}
+
+// openAPIV2ToDefinition builds a schemaDefinition for the given schemaID based on
+// Resource information from OpenAPI v2 endpoint
+func openAPIV2ToDefinition(models proto.Models, modelName string, version string) (schemaDefinition, error) {
+	protoSchema := models.LookupModel(modelName)
+	switch m := protoSchema.(type) {
+	case *proto.Map:
+		// If the schema is a *proto.Map, it will not have any Fields associated with it
+		// even though all Kubernetes resources have at least apiVersion, kind and metadata.
+		//
+		// We transform this Map to a Kind and inject these fields from
+		// a known existing Kubernetes resource (ConfigMap).
+		configMap := models.LookupModel("io.k8s.api.core.v1.ConfigMap")
+		apiVersion := configMap.(*proto.Kind).Fields["apiVersion"]
+		apiVersion.(*proto.Primitive).Path = m.Path.FieldPath("apiVersion")
+		kind := configMap.(*proto.Kind).Fields["kind"]
+		kind.(*proto.Primitive).Path = m.Path.FieldPath("kind")
+		metadata := configMap.(*proto.Kind).Fields["metadata"]
+		metadata.(*proto.Ref).Path = m.Path.FieldPath("metadata")
+		protoSchema = &proto.Kind{
+			BaseSchema: m.BaseSchema,
+			Fields: map[string]proto.Schema{
+				"apiVersion": apiVersion,
+				"kind":       kind,
+				"metadata":   metadata,
+			},
+		}
```

Comment:
> The Ref type may have some internal fields that are hard to set, but I would look into implementing the proto.Reference interface in a custom type. It looks like that is pretty simple (string and a sub-schema) - should be reasonable.

Nice, I couldn't find an easy way creating a ref. Let me take a look at the proto.Reference interface.

> if I request a TestObj which has a field UserAttribute, then the Userattribute that this retrieves won't have these values.

I don't believe this is supported by the CRDs spec. There is still an open issue about this: https://github.com/kubernetes/kubernetes/issues/62872, which seems to be what your example describes.

Reply:
Implementing `proto.Reference` worked fine.

I've removed the dependency on ConfigMap. I've moved the logic to its own function. I have not added any test specifically on this new functions, I'm thinking of expanding the tests for openAPIV2ToDefinition instead which makes use of this, just in case the new functions get changed again. Let me know what you think.